delta-app/src/main/java/io/cdap/delta/app/DeltaWorker.java

```diff
@@ -17,133 +17,284 @@
-    eventConsumer = target.createConsumer(deltaContext);
-    readerDefinition = GSON.fromJson(context.getSpecification().getProperty(READER_DEFINITION),
-                                     EventReaderDefinition.class);
     OffsetAndSequence offsetAndSequence = deltaContext.loadOffset();
     offset = offsetAndSequence.getOffset();
-    DirectEventEmitter emitter = new DirectEventEmitter(eventConsumer, deltaContext,
-                                                        offsetAndSequence.getSequenceNumber(), readerDefinition);
-    eventReader = source.createReader(readerDefinition, deltaContext, emitter);
+
+    Set<SourceTable> expandedTables = config.getTables().stream()
+      .map(t -> {
+        Set<DMLOperation> expandedDmlBlacklist = new HashSet<>(t.getDmlBlacklist());
+        expandedDmlBlacklist.addAll(config.getDmlBlacklist());
+        Set<DDLOperation> expandedDdlBlacklist = new HashSet<>(t.getDdlBlacklist());
+        expandedDdlBlacklist.addAll(config.getDdlBlacklist());
+        return new SourceTable(t.getDatabase(), t.getTable(), t.getSchema(),
+                               t.getColumns(), expandedDmlBlacklist, expandedDdlBlacklist);
+      }).collect(Collectors.toSet());
+    readerDefinition = new EventReaderDefinition(expandedTables,
+                                                 config.getDmlBlacklist(),
+                                                 config.getDdlBlacklist());
+
+    // TODO: make the queue size configurable?
+    eventQueue = new ArrayBlockingQueue<>(100);
   }
 
   @Override
   public void run() {
     if (offset.get().isEmpty()) {
       deltaContext.setOK();
     }
-    eventConsumer.start();
-    eventReader.start(offset);
     try {
-      stopLatch.await();
-    } catch (InterruptedException e) {
-      // ignore and return, this is just waiting to stop anyway
+      startFromLastCommit();
+    } catch (Exception e) {
+      // if this fails at the start of the run, fail the entire run as it probably means there is something
+      // wrong with the configuration or environment
+      throw new RuntimeException(e.getMessage(), e);
+    }
+
+    AtomicBoolean interrupted = new AtomicBoolean(false);
+    // in most circumstances, the worker is stopped, which causes eventQueue.take() to be interrupted and break out
+    // of the loop. However, it is also possible for the worker to be stopped before it gets here, in which case it
+    // should just end immediately.
+    while (!shouldStop) {
+
+      // create the retry policy for handling exceptions thrown from calls to applyEvent()
+      // the worker will retry up to the configured maximum amount of time.
+      // On each failed attempt, offset will be rolled back to the last commit, and any in-memory state
+      // related to metrics and queued change events will be cleared away.
+      RetryPolicy<Object> retryPolicy = new RetryPolicy<>()
+        .withMaxAttempts(Integer.MAX_VALUE)
```

Comment:
Seems we are going to be capped by max duration, do we have to provide max attempts?

Reply:
i'll add a comment, but if you don't set this, it defaults to 3 max attempts